File: skills/render-create/templates/background-tasks/py/workflow.py
"""
Render Workflow

Uses the Render SDK to define distributed tasks.
See: https://pypi.org/project/render-sdk/
"""

import asyncio
import logging

from dotenv import load_dotenv
from render_sdk import Retry, Workflows

load_dotenv()

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Retry configuration
retry = Retry(max_retries=3, wait_duration_ms=1000, backoff_scaling=1.5)

# Initialize Workflows app
app = Workflows(
    default_retry=retry,
    default_timeout=300,
    auto_start=True,
)


@app.task
def square(a: int) -> int:
    """Square a number (subtask)."""
    logger.info(f"Computing square of {a}")
    return a * a


@app.task
async def add_squares(a: int, b: int) -> int:
    """Add the squares of two numbers using subtasks."""
    logger.info(f"Computing add_squares: {a}, {b}")

    result1 = await square(a)
    result2 = await square(b)

    total = result1 + result2
    logger.info(f"Total: {total}")
    return total


@app.task
async def process_data(data: str) -> str:
    """Process data."""
    logger.info(f"Processing data: {data}")

    # TODO: Add your data processing logic here
    result = data.upper()

    logger.info(f"Processed result: {result}")
    return result


@app.task
async def fan_out(items: list[str]) -> list[str]:
    """Process multiple items in parallel using subtasks."""
    logger.info(f"Processing {len(items)} items in parallel")

    tasks = [process_data(item) for item in items]
    results = await asyncio.gather(*tasks)

    return list(results)


# Task server starts automatically with auto_start=True

File: skills/render-create/templates/background-tasks/ts/workflow.ts
/**
 * Render Workflow
 *
 * Uses the Render SDK to define distributed tasks.
 * See: https://github.com/render-oss/sdk
 */

import "dotenv/config";
import { task, type Retry } from "@renderinc/sdk/workflows";

// Retry configuration
const retry: Retry = {
  maxRetries: 3,
  waitDurationMs: 1000,
  factor: 1.5,
};

/**
 * Simple task that squares a number (subtask)
 */
const square = task({ name: "square" }, function square(a: number): number {
  console.log(`Calculating square of ${a}`);
  return a * a;
});

/**
 * Async task that adds two squared numbers
 */
task(
  {
    name: "addSquares",
    timeoutSeconds: 300,
    retry,
  },
  async function addSquares(a: number, b: number): Promise<number> {
    console.log(`Adding squares of ${a} and ${b}`);

    const result1 = await square(a);
    const result2 = await square(b);

    const sum = result1 + result2;
    console.log(`Result: ${result1} + ${result2} = ${sum}`);
    return sum;
  },
);

/**
 * Task that processes data
 */
task(
  {
    name: "processData",
    timeoutSeconds: 300,
    retry,
  },
  async function processData(data: string): Promise<string> {
    console.log(`Processing data: ${data}`);

    // TODO: Add your data processing logic here
    const result = data.toUpperCase();

    console.log(`Processed result: ${result}`);
    return result;
  },
);

// Task server starts automatically when RENDER_SDK_SOCKET_PATH is set
